monitoring/mongodb/alerts.yaml

```diff
@@ -97,20 +103,53 @@ groups:
   - alert: ReplicationLagWarning
     expr: |
       label_replace(
-        max(
-            max(mongodb_rs_members_optimeDate{namespace="${namespace}",pod=~"${service}.*",member_state="PRIMARY"})
-                by(pod, rs_nm)
-            - ignoring(member_idx) group_right
-            min(mongodb_rs_members_optimeDate{namespace="${namespace}",pod=~"${service}.*",member_state="SECONDARY"})
-                by(pod, rs_nm, member_idx)
-        ) by(member_idx, rs_nm) / 1000 > 30
-      , "member_idx", "$1", "member_idx", "(${service}[^.]*)\\.${service}.*")
-    for: 10m
+        (
+          scalar(max(mongodb_rs_members_optimeDate{namespace="${namespace}",pod=~"${service}.*",member_state="PRIMARY"}))
+          - max by(member_idx) (mongodb_rs_members_optimeDate{namespace="${namespace}",pod=~"${service}.*",member_state="SECONDARY"})
+        ) / scalar(
+          max(
+            (mongodb_mongod_replset_oplog_head_timestamp{namespace="${namespace}",pod=~"${service}.*"}
+            - on(pod)
+            mongodb_mongod_replset_oplog_tail_timestamp{namespace="${namespace}",pod=~"${service}.*"})
+            * on(pod) group_left()
+            (mongodb_mongod_replset_my_state{namespace="${namespace}",pod=~"${service}.*"} == 1)
+          )
+        ) * 100 > ${replicationLagWarningThreshold},
+        "member_idx", "$1", "member_idx", "(${service}[^.]*)\\.${service}.*"
+      )
+    for: 8m
```

Comment:
Approach looks good to me, however these metrics return values for both of the `shardsrv` and `configsrv` instances. Is there a risk of mix between their values in the calculation that may trigger a false alert or the other way arround by not triggering the alert ?

I'm saying this because of the `max()` operators we use in multiple parts of the query and that remove the labels.

Note: `zenko-operator` replaces the `service` variable with `data-db-mongodb-sharded` which is the prefix of both the `shardsrv` and `configsrv` instances.

Reply:
Practically this won't affect much but indeed it might create some <0 metrics sometimes. the solution to this is to rely on the `rs_nm`, as the config server is using a dedicated replicaset, ensuring metrics are independent. Last commit was updated.